### ISSUE TITLE
Linkage Error in Tomcat using multiple webapp instances 

### ISSUE THREAD

**Issue by sancai** [skadistats/clarity#38]
Good day, fellow devs,

First of all, this might not be a clarity issue.

I've been cracking my head on this one for a couple of days:
When using clarity in a Tomcat 7 container which have multiple webapps deployed I'm getting this error:
```
...
Caused by: java.lang.LinkageError: 
loader constraint violation: when resolving method "skadistats.clarity.wire.proto.Demo$CDemoFileInfo.parseFrom(Lcom/google/protobuf/ByteString;)Lskadistats/clarity/wire/proto/Demo$CDemoFileInfo;" 
the class loader (instance of <bootloader>) 
of the current class, java/lang/Object, 
and the class loader (instance of org/apache/catalina/loader/WebappClassLoader) 
for resolved class, skadistats/clarity/wire/proto/Demo$CDemoFileInfo, 
have different Class objects for the type .parseFrom(Lcom/google/protobuf/ByteString;)Lskadistats/clarity/wire/proto/Demo$CDemoFileInfo; 
used in the signature
	at java.lang.invoke.MemberName$Factory.resolve(MemberName.java:847)
	at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:872)
```
Some background: we have the production and the staging environment located on the same physical server, in the same Tomcat container deployed as different webapps.
The error occures in whichever environment tries to load clarity classes for the **second** time.
There's no difference in staging/production environment in terms of code/packages used, but there are some differences in configuration: db instance used, logging, etc.. The clarity jar is included in each webapps' WEB-INF/lib folder. Turning on classloader debugging shows the it tries to load the class from 2 respective Document Root.

After some research it seems that this type of error is usually related to multiple different versions of the same package (fetched by dependencies) referenced by the webapp. This case is different. I'm using the same clarity version in different webapps.

I could have copied clarity jar to Tomcat's global lib folder, this way i can avoid having 2 files to load the classes from. I have 2 problems with this approach:
- this file should be updated when clarity is updated. In other words, it would be seperated from application code which I would like to avoid
- there are literally myriad of other classes that exist in these 2 webapps with the same version and signature and they don't produce linkage errors 
 
I'm not an expert in Java classloader issues, so thought I give it a try, perhaps someone here know better. What am I missing?

**Comment by spheenik** [comment]
Really strange....
First thing I do not understand is: How is `.parseFrom(Lcom/google/protobuf/ByteString;)Lskadistats/clarity/wire/proto/Demo$CDemoFileInfo` a type? I would say it's a method?

**Comment by sancai** [comment]
Exactly.. To be honest I'm running out of plans how to fix this. It just
doesn't make sense. Probably I'll fire up 2 different Tomcat instances to
handle it.
Martin, let's keep this issue open for a couple of days for visibility
after that I'll close it since i don't really think it is related to
clarity.

**Comment by spheenik** [comment]
Unfortunately I am away for a week now and cannot investigate.

After that I'll try to find some time to look at it. Maybe the thing it complains about is some kind of inner class?
Btw: Is this clarity 1 or 2?

Anyway, please leave open until further notice. thx!

**Comment by lmdg** [comment]
Hi guys, I had a similar [issue recently](http://stackoverflow.com/q/30703738/1626062) and had stumbled upon your issue while searching. I was able to resolve my issue by updating the JDK used (making sure it was at least 7u55. Just thought I'd let you know since it might help with your issue.

**Comment by sancai** [comment]
Hi lmdg,

Thanks for the heads up.
Unfortunately this we're already on 7u80 and this is the last update for
java 7.

**Comment by spheenik** [comment]
Ok, I have looked into it and I kind of know what is going on:

The problem lies in this function:
https://github.com/skadistats/clarity-protobuf/blob/master/src/main/java/skadistats/clarity/wire/PacketTypes.java#L172-L189

Given a subclass of GeneratedMessage (Protobuf) it's supposed to return a MethodHandle that can be invoked to parse a ByteString into a GeneratedMessage.

I *think* it complains because the MethodHandle is shared between class loaders, and it's return type (CDemoFileInfo) is different in each class loader.

I guess you could call this an inconsistency (a bug) in the JDK implementation. I don't how to proceed.

What you could do is copy the whole PacketTypes.java over to your project (leaving the package the same) and start experimenting :)

**Comment by spheenik** [comment]
I made this to squeeze out some milliseconds. I'll revert it to the old way using basic reflection, probably.

**Comment by sancai** [comment]
Thanks for the efforts Martin. I'm trying to get my head around what you
have written. I'm not that familiar with dynamic method invokation.
Can it be the issue, that
MethodHandles.publicLookup().findStatic

will return a MethodHandle from a different jar/package ?
Anyway, right now I don't have the time to play with it, but I'll probably
find some time in 1-2 weeks.

Regards,
//S

**Comment by spheenik** [comment]
No problem, if you do not understand or have time, I will fix it soon. (When importing S2 protobufs I will build a new version anyway :)

> Am 16.06.2015 um 12:04 schrieb sancai <notifications@github.com>:
> 
> Thanks for the efforts Martin. I'm trying to get my head around what you
> have written. I'm not that familiar with dynamic method invokation.
> Can it be the issue, that
> MethodHandles.publicLookup().findStatic
> 
> will return a MethodHandle from a different jar/package ?
> Anyway, right now I don't have the time to play with it, but I'll probably
> find some time in 1-2 weeks.
> 
> Regards,
> //S
> 
> 
> 
> 
> 
>

**Comment by spheenik** [comment]
I reverted back to using normal reflection instead of MethodHandles, so this should not happen anymore.
Fixed in 1.3 and 2.0+
